src/index/Vocabulary.h

```diff
@@ -70,9 +72,27 @@ class Vocabulary {
   //! Append a word to the vocabulary. Wraps the std::vector method.
   void push_back(const string& word) { _words.push_back(word); }
 
-  //! Get the word with the given id (as const, not as lvalue).
-  const string& operator[](Id id) const {
-    return _words[static_cast<size_t>(id)];
+  //! Get the word with the given id or an empty optional if the
+  //! word is not in the vocabulary.
+  //! Note: So as to keep things as references this does *NOT* check the
+  //! externalized part of the vocabulary
+  const std::optional<std::reference_wrapper<const string>> operator[](
+      Id id) const {
+    if (id < _words.size()) {
+      return _words[static_cast<size_t>(id)];
+    } else {
+      return std::nullopt;
+    }
+  }
```

Comment:
I never thought about this, but now that I see it i don't like it:
We have to know, that with the externalized vocabulary (knowledge base vocab), we must NOT call operator[] from the outside but idToOptionalString but, if we know that nothing is externalized, we can call the operator[]. I think this is a recipe for desaster. I can agree to merge this for now, since it is not the fault of your PR but I suggest the following with my PrefixCompression:
There we either have to always check for `_words.size()` or we completely disable the operator[] since we only can return strings because of the compression. And for uncompressed vocabularies (text vocab) we forbid externalization, so the operator[] or "idToOptionalStringRef" always works as expected.

Reply:
Another argument: this is a really unexpected interface, if operator[] returns an optional.

Reply:
Ok then I'd prefer to disable `operator[]` in your PR

Reply:
Agreed, we only leave idToOptionalStringRef (if uncompressed -> unexternalized) and idToOptionalString (if compressed -> possibly externalized).